### DISCUSSION TITLE
Allow to encrypt columns in DB

### DISCUSSION THREAD

**Discussion by miry** [martenframework/marten#238] [Ideas]
I would like to migrate Rails application to Marten.
One of the features that Active Record has is column encryption.

What do you think if it makes sense to have this feature in Marten?

## Reference

[Active Record Encryption](https://guides.rubyonrails.org/active_record_encryption.html)

**Comment by ellmetha** [comment]
Hey! 👋

I never had to leverage Active Record Encryption myself but I think that this is definitely a feature that could make its way to the framework at one point. I am willing to explore the question and create a feature issue for it. To be honest, I don't think this is something that will be a high priority for the framework in the short term but this is something that could be added further down the line. I will also gladly review PRs that would add this feature sooner if someone outside of the Core team wants to take a stab at adding this. 🙂 

In the meantime, would handling encryption "manually" in your application be a viable option for you (for example, by leveraging the built-in [`Marten::Core::Encryptor`](https://martenframework.com/docs/api/0.4/Marten/Core/Encryptor.html) helper)?

**Reply by miry** [reply]
Yep, I came to Encryptor helper and that why found it is very helpful for this feature. What is missing for now to have a contribution, how I can update Model's field to use some sort of Decorator to encode/decode field on the fly. If it is already in place, then it would be easy.

Something like this:

```
field :email, String, read: -> {|a| decrypt(a) },  write: -> {|a| encrypt(a) }
```


May be there is a custom Type, that would be the type of the field?

```
field :email, EncryptString
```

**Reply by ellmetha** [reply]
What about leveraging [model callbacks](https://martenframework.com/docs/models-and-databases/callbacks) to automatically encrypt/decrypt the right model field values? For example, [`after_initialize`](https://martenframework.com/docs/models-and-databases/callbacks#after_initialize) could be used for decrypting values while [`before_save`](https://martenframework.com/docs/models-and-databases/callbacks#before_save-and-after_save) could be used for encrypting values (before saving records). This is just an example, there might be other options.
